### PR TITLE
feat: any to near advanced logic

### PR DESCRIPTION
**Pull request by leoslr**
- Add the advanced logic extensions for payment with conversion for native token.
- Add the native token near specific case
- Add the associated tests

### CODE REVIEW

**Review by yomarion — COMMENTED** [review]
Seems good but I think there are a few maintenability and readability improvements

File: packages/types/src/extension-types.ts [hunk 1]
```diff
@@ -78,6 +78,7 @@ export enum ID {
   PAYMENT_NETWORK_ETH_FEE_PROXY_CONTRACT = 'pn-eth-fee-proxy-contract',
   PAYMENT_NETWORK_ETH_INPUT_DATA = 'pn-eth-input-data',
   PAYMENT_NETWORK_NATIVE_TOKEN = 'pn-native-token',
+  PAYMENT_NETWORK_ANY_TO_NATIVE_TOKEN = 'pn-any-to-native-token',
```
Comment (kevindavee) [inline packages/types/src/extension-types.ts:81]: So this one will replace `pn-eth-input-data`?

File: packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+          salt,
+          network: 'aurora',
+        }),
+      ).toBeTruthy();
+    });
+    it('createCreationAction() throws with unsupported payment network', () => {
+      expect(() => {
+        new AnyToNearPaymentNetwork(currencyManager).createCreationAction({
+          ...partialCreationParams,
+          paymentNetworkName: 'another-chain',
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts:255]: I think this is a mistake, I removed the `paymentNetworkName` variable everywhere except here it appears.
It's a parameter from a Payment Network we extend of. It it useull in the said PN but useless here as we already have the network parameter.
Using both of them would only complicate things.

File: packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+
+  describe('action creations, edge cases', () => {
+    const partialCreationParams: ExtensionTypes.PnAnyToAnyConversion.ICreationParameters = {
+      salt,
+      refundAddress: 'refund.near',
+      feeAddress: 'fee.near',
+      feeAmount: '100',
+      maxRateTimespan: 1000000,
+    };
+    it('createCreationAction() works with no payment, refund address, feeAddress and feeAmount', () => {
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts:243]: Indeed it's the same, I got them from different tests files so there might be some overlap

File: packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+            invalidRequestState,
+            payeeRaw.identity,
+            arbitraryTimestamp,
+          ),
+        ).toThrowError(
+          'The currency (EUR) of the request is not supported for this payment network',
+        );
+      });
+      it('throws when network is undefined', () => {
+        creationAction.parameters.network = undefined;
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/test/extensions/payment-network/any-to-near.test.ts:312]: Before each test I already do `creationAction = utils.deepCopy(actionCreationWithAnyToNativeTokenPayment);`
So the creation action is a new 'clean' every time.

File: packages/types/src/extension-types.ts (hunk 1 above)
Comment (leoslr) [reply packages/types/src/extension-types.ts:81]: @kevindavee the `pn-any-to-eth-proxy` is specific to ethereum (or evm-chain) while the `any-to-native` one can easily be updated to adapt to the native currency of any chain.

In the future it might be a good idea to make the `any-to-eth-proxy` an extension of the `any-to-native` for more consistency.

**Review by alexandre-abrioux — COMMENTED** [review]
Code is great! I'm just missing some context:
- What is this PN adding that [any-to-eth-proxy](https://github.com/RequestNetwork/requestNetwork/blob/master/packages/advanced-logic/specs/payment-network-any-to-eth-proxy-0.1.0.md) does not support, apart from the fact that this is not for EVM?
- What was the reason for adding this PN? (what feature is missing)
- Maybe add the [specs](https://github.com/RequestNetwork/requestNetwork/tree/master/packages/advanced-logic/specs) for any-to-native to clarify

File: packages/advanced-logic/src/advanced-logic.ts [hunk 5]
```diff
@@ -124,6 +130,11 @@ export default class AdvancedLogic implements AdvancedLogicTypes.IAdvancedLogic
           `extension with id: ${id} not found for network: ${requestState.currency.network}`,
         );
       }
+
+      if (id === ExtensionTypes.ID.PAYMENT_NETWORK_ANY_TO_NATIVE_TOKEN) {
```
Comment (yomarion) [inline packages/advanced-logic/src/advanced-logic.ts:134]: I think this can be slightly refactored into a more elegant way (check the block before)

File: packages/advanced-logic/src/extensions/payment-network/any-to-near.ts [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    extensionAction: ExtensionTypes.IAction,
+  ): void {
+    const network =
+      extensionAction.action === ExtensionTypes.PnFeeReferenceBased.ACTION.CREATE
+        ? extensionAction.parameters.network
+        : request.extensions[this.extensionId]?.values.network;
+
+    if (!network) {
+      throw new Error(
+        `The network must be provided from the creation action or from the extension state`,
```
Comment (yomarion) [inline packages/advanced-logic/src/extensions/payment-network/any-to-near.ts:193]:
```suggestion
        `The network must be provided by the creation action or by the extension state`,
```

**Review by yomarion — APPROVED** [review]
Good job @leoslr

File: packages/advanced-logic/src/extensions/payment-network/any-to-native.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  public createCreationAction(
+    creationParameters: ExtensionTypes.PnAnyToAnyConversion.ICreationParameters,
+  ): ExtensionTypes.IAction<ExtensionTypes.PnAnyToAnyConversion.ICreationParameters> {
+    const network = creationParameters.network;
+    if (!network) {
+      throw Error('network is required');
+    }
+
+    this.throwIfInvalidNetwork(network, this.supportedNetworks);
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/src/extensions/payment-network/any-to-native.ts:22]: If I remember correctly, I did not integrate it to the `throwIfInvalidNetwork` function because then the `network` variable would still be considered as possibly undefined which raised errors.
I've made an update so the check returns the `network` property as a string if there's no error thrown.

File: packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts [hunk 8]
```diff
@@ -148,7 +148,7 @@ describe('extensions/payment-network/erc20/any-to-erc20-fee-proxy-contract', ()
           network: 'kovan',
           acceptedTokens: ['0x0000000000000000000000000000000000000003'],
         });
-      }).toThrowError('network kovan not supported');
+      }).toThrowError(`Payment network 'kovan' is not supported by this extension (only`);
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/test/extensions/payment-network/any-to-erc20-proxy.test.ts:151]: The error state the list of supported networks, I've added it the the expected error message

**Comment by leoslr** [comment]
> Code is great! I'm just missing some context:
> 
> * What is this PN adding that [any-to-eth-proxy](https://github.com/RequestNetwork/requestNetwork/blob/master/packages/advanced-logic/specs/payment-network-any-to-eth-proxy-0.1.0.md) does not support, apart from the fact that this is not for EVM?
> * What was the reason for adding this PN? (what feature is missing)
> * Maybe add the [specs](https://github.com/RequestNetwork/requestNetwork/tree/master/packages/advanced-logic/specs) for any-to-native to clarify

@alexandre-abrioux 
- any-to-eth was EVM specific, this adds a generic PN (any-to-native) that can be used as a base to create other PN - any-to-near in this case, but we could also base any-to-eth on it.
- For supporting any-to-near, we needed a new PN since some checks are network related, like the address validation.
- I'll have to update the specs yes

**Comment by leoslr** [comment]
@yomarion 
I've made an update based on Alex comments that checks that `network` is define within the `throwIfInvalidNetwork` function.
It makes this test fail https://github.com/RequestNetwork/requestNetwork/blob/master/packages/advanced-logic/test/extensions/payment-network/native-token.test.ts#L177
Are there any reason we want this to work ?

File: packages/advanced-logic/src/extensions/payment-network/any-to-native.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexandre-abrioux) [reply packages/advanced-logic/src/extensions/payment-network/any-to-native.ts]: To resolve this you could do:
```ts
  protected throwIfInvalidNetwork(network?: string): asserts network is string {
    if (!network) {
      throw Error('network is required');
    }
    if (network && this.supportedNetworks && !this.supportedNetworks.includes(network)) {
      throw new UnsupportedNetworkError(network, this.supportedNetworks);
    }
  }
```

**Review by alexandre-abrioux — APPROVED** [review]

File: packages/advanced-logic/src/extensions/payment-network/any-to-native.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/src/extensions/payment-network/any-to-native.ts]: @alexandre-abrioux nice, I did not know about this TS feature, thanks !

**Comment by yomarion** [comment]
> @yomarion I've made an update based on Alex comments that checks that `network` is define within the `throwIfInvalidNetwork` function. It makes this test fail https://github.com/RequestNetwork/requestNetwork/blob/master/packages/advanced-logic/test/extensions/payment-network/native-token.test.ts#L177 Are there any reason behind this test ? It looks a bit weird to me.

I think it's to cover this case:
- I create a request with a currency & network, without advanced-logic
- I add an extension to the advanced logic, with a createAction. This action does not have to have a network, we get it from the request

File: packages/advanced-logic/src/extensions/payment-network/any-to-near.ts [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+    if (!supportedNetworks.includes(network)) {
+      throw new Error(`The network (${network}) is not supported for this payment network.`);
+    }
+
+    const currency = this.currencyManager.fromStorageCurrency(request.currency);
+    if (!currency) {
+      throw new UnsupportedCurrencyError(request.currency);
+    }
+    if (!this.currencyManager.supportsConversion(currency, network) && currency.id !== 'USD') {
```
[the remark this answers is not in the document]
Comment (leoslr) [reply packages/advanced-logic/src/extensions/payment-network/any-to-near.ts]: I Removed the USD dependency

**Comment by leoslr** [comment]
> > @yomarion I've made an update based on Alex comments that checks that `network` is define within the `throwIfInvalidNetwork` function. It makes this test fail https://github.com/RequestNetwork/requestNetwork/blob/master/packages/advanced-logic/test/extensions/payment-network/native-token.test.ts#L177 Are there any reason behind this test ? It looks a bit weird to me.
> 
> I think it's to cover this case:
> 
> * I create a request with a currency & network, without advanced-logic
> * I add an extension to the advanced logic, with a createAction. This action does not have to have a network, we get it from the request

Note:
We discussed this with Yoann and decided to go forward and remove the test
